source/common/router/router.cc

```diff
@@ -621,6 +594,65 @@ void Filter::onUpstreamReset(UpstreamResetType type,
   }
 }
 
+bool Filter::maybeRetryReset(const absl::optional<Http::StreamResetReason>& reset_reason) {
+  Upstream::HostDescriptionConstSharedPtr upstream_host;
+  if (upstream_request_) {
+    upstream_host = upstream_request_->upstream_host_;
+  }
+
+  // We don't retry on a global timeout or if we already started the response.
+  if (!downstream_response_started_ && retry_state_) {
+    // Notify retry modifiers about the attempted host.
+    if (upstream_host != nullptr) {
+      retry_state_->onHostAttempted(upstream_host);
+    }
+
+    RetryStatus retry_status =
+        retry_state_->shouldRetry(nullptr, reset_reason, [this]() -> void { doRetry(); });
+    if (retry_status == RetryStatus::Yes && setupRetry(true)) {
+      if (upstream_host) {
+        upstream_host->stats().rq_error_.inc();
+      }
+      return true;
+    } else if (retry_status == RetryStatus::NoOverflow) {
+      callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::UpstreamOverflow);
+    } else if (retry_status == RetryStatus::NoRetryLimitExceeded) {
+      callbacks_->streamInfo().setResponseFlag(
+          StreamInfo::ResponseFlag::UpstreamRetryLimitExceeded);
+    }
+  }
+
+  return false;
+}
+
+void Filter::onUpstreamReset(const absl::optional<Http::StreamResetReason>& reset_reason) {
```

Comment:
i'm basically redefining what "upstream reset" means here. it used to mean a reset by the upstream, or a local reset of the upstream request (timeout). Now it only means the upstream reset it because we handle global/per try timeouts in other methods